rsocket-core/src/main/java/io/rsocket/resume/InMemoryResumableFramesStore.java

```diff
@@ -19,118 +19,240 @@
+  final Queue<ByteBuf> cachedFrames;
   final String tag;
   final int cacheLimit;
 
   volatile long impliedPosition;
   static final AtomicLongFieldUpdater<InMemoryResumableFramesStore> IMPLIED_POSITION =
       AtomicLongFieldUpdater.newUpdater(InMemoryResumableFramesStore.class, "impliedPosition");
 
-  volatile long position;
-  static final AtomicLongFieldUpdater<InMemoryResumableFramesStore> POSITION =
-      AtomicLongFieldUpdater.newUpdater(InMemoryResumableFramesStore.class, "position");
+  volatile long firstAvailableFramePosition;
+  static final AtomicLongFieldUpdater<InMemoryResumableFramesStore> FIRST_AVAILABLE_FRAME_POSITION =
+      AtomicLongFieldUpdater.newUpdater(
+          InMemoryResumableFramesStore.class, "firstAvailableFramePosition");
 
-  volatile int cacheSize;
-  static final AtomicIntegerFieldUpdater<InMemoryResumableFramesStore> CACHE_SIZE =
-      AtomicIntegerFieldUpdater.newUpdater(InMemoryResumableFramesStore.class, "cacheSize");
+  long remoteImpliedPosition;
 
-  CoreSubscriber<? super Void> saveFramesSubscriber;
+  int cacheSize;
 
-  CoreSubscriber<? super ByteBuf> actual;
+  Throwable terminal;
 
-  /**
-   * Indicates whether there is an active connection or not.
-   *
-   * <ul>
-   *   <li>0 - no active connection
-   *   <li>1 - active connection
-   *   <li>2 - disposed
-   * </ul>
-   *
-   * <pre>
-   * 0 <-----> 1
-   * |         |
-   * +--> 2 <--+
-   * </pre>
-   */
-  volatile int state;
-
-  static final AtomicIntegerFieldUpdater<InMemoryResumableFramesStore> STATE =
-      AtomicIntegerFieldUpdater.newUpdater(InMemoryResumableFramesStore.class, "state");
+  CoreSubscriber<? super ByteBuf> actual;
+  CoreSubscriber<? super ByteBuf> pendingActual;
+
+  volatile long state;
+  static final AtomicLongFieldUpdater<InMemoryResumableFramesStore> STATE =
+      AtomicLongFieldUpdater.newUpdater(InMemoryResumableFramesStore.class, "state");
+
+  static final long FLAG_FINALIZED =
+      0b1000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000L;
+  static final long FLAG_DISPOSED =
+      0b0100_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000L;
+  static final long FLAG_TERMINATED =
+      0b0010_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000L;
+  static final long FLAG_CONNECTED =
+      0b0001_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000L;
+  static final long FLAG_PENDING_CONNECTION =
+      0b0000_1000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000L;
+  static final long FLAG_REMOTE_IMPLIED_POSITION_CHANGED =
+      0b0000_0100_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000L;
+  static final long FLAG_HAS_FRAME =
+      0b0000_0010_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000L;
+  static final long MAX_WORK_IN_PROGRESS =
+      0b0000_0000_0000_0000_0000_0000_0000_0000_1111_1111_1111_1111_1111_1111_1111_1111L;
```

Comment:
A brief explanation of each state and a diagram would serve a good purpose and should be relatively easy to add.

Reply:
fixed